src/librbd/image/OpenRequest.cc

```diff
@@ -476,10 +477,14 @@ Context *OpenRequest<I>::handle_v2_get_data_pool(int *result) {
     *result = util::create_ioctx(m_image_ctx->md_ctx, "data pool", data_pool_id,
                                  {}, &m_image_ctx->data_ctx);
     if (*result < 0) {
-      send_close_image(*result);
-      return nullptr;
+      if (!m_ignore_data_pool_error) {
+        send_close_image(*result);
+        return nullptr;
+      }
+      m_image_ctx->invalid_data_pool = true;
```

Comment:
Nit: perhaps also call `m_image_ctx->data_ctx.close()` to ensure it cannot be incorrectly used. In a perfect world `librados::IoCtx` would also provide a `operator unspecified_bool_type() const`-like helper [1] to see if the `IoCtx` is valid -- and then you could eliminate that extra bool in `ImageCtx`.

[1] https://beta.boost.org/doc/libs/1_48_0/boost/smart_ptr/detail/operator_bool.hpp

Reply:
So, do you prefer if I added "safe bool" operator to `librados::IoCtx` in this PR and avoided adding extra bool in `ImageCtx`? I can try this. BTW, do I understand this right that since  C++11 we could just use `explicit operator bool()` instead, but we can't do this here because it is a part of API?

Reply:
Or I could just add `bool librados::IoCtx::is_valid() const`, which is not so nice as "safe bool", but easier to provide as API.

Reply:
A `is_valid` method is fine w/ me as well.